benchmark/C/Savina/src/parallelism/GuidedSearch.lf

```diff
@@ -0,0 +1,246 @@
+ * compared to single threaded execution. This is likely due to the race-conditions
+ * mentioned before.
+ *  
+ * @author Christian Menard
+ * @author Hannes Klein
+ * @author Soroush Bateni
+ */
+
+target CCpp {
+    /* [[[cog
+      if (threaded_runtime=="True"):
+          cog.outl(f"threads: {threads},")
+      else:
+          cog.outl("threads: 0,")
+    ]]] */
+    threads: 0,
+    /// [[[end]]]
+    build-type : RelWithDebInfo,
+    cmake-include: "GuidedSearch/GuidedSearch.cmake",
+    files: ["GuidedSearch/GuidedSearchCommon.cc", 
+        "GuidedSearch/GuidedSearchCommon.hh", 
+        "GuidedSearch/GridNode.cc", 
+        "GuidedSearch/GridNode.hh"],
+    logging: "warn"
+};
+
+preamble {=
+    #include "GuidedSearchCommon.hh"
+    #include <deque>
+    #include <list>
+    #include <iostream>
+=}
+
+reactor Manager(numWorkers: size_t(20), gridSize: size_t(30), priorities: size_t(30)) {
+
+    state grid: Grid*;
+    state nodesToVisit: std::deque<GridNode*>;
+    state pathWasFound: bool(false);
+    
+    input start: bool;
+    output finished: bool;
+    
+    output[numWorkers] search: WorkMessage;
+    input[numWorkers] pathFound: bool;
+    input[numWorkers] moreNodesToVisit: std::deque<GridNode*>;
+    
+    logical action next;
+    
+    reaction (start) -> next {=
+        self->grid = new Grid(self->gridSize, self->priorities);
+        self->nodesToVisit = std::deque<GridNode*>();
+        srand(123456);
+        self->grid->initializeData(); // reset the self->grid
+        self->nodesToVisit.clear();
+        self->pathWasFound = false;
+        
+        // start execution, first visit the origin node
+        self->nodesToVisit.push_back(self->grid->originNode());
+        schedule(next, 0);
+    =}
+    
+    reaction (next) -> next, search {=
+        // abort if path was found
+        if (self->pathWasFound) {
+            return;
+        }
+
+        // send nodes from the internal queue to all workers
+        size_t workerIndex{0};        
+        while(!self->nodesToVisit.empty() && workerIndex < self->numWorkers) {
+            GridNode* startNode = self->nodesToVisit.front();
+            self->nodesToVisit.pop_front();
+            SET(search[workerIndex], WorkMessage(startNode, self->grid->targetNode()));
+            workerIndex++;
+        }
+        
+         // continue with the next iteration
+         schedule(next, 0);
+    =}
+    
+    reaction(pathFound) -> finished {=
+        self->pathWasFound = true;
+        bool valid = self->grid->validate();
+        info_print("Found a path through the grid");
+        info_print("Result valid: %d", valid);
+        SET(finished, true);
+    =}
+    
+    reaction (moreNodesToVisit) {=
+        // collect more nodes to visit from all workers and store them in the internal queue
+        for (int i=0; i< moreNodesToVisit_width; i++) {
+            if (moreNodesToVisit[i]->is_present) {
+                auto nodes = moreNodesToVisit[i]->value;
```

Comment:
This copies the entire deque from `moreNodesToVisit[i]->value` to `node`. Is there a reason why you removed the const reference from the original code?

Reply:
I changed this back to be a const reference.